Cache Link Checking of Lychee (markdown files)

/kind feature

When we test a lot via CI, then we hit a rate-limit any Lychee will fail:

```
lychee --verbose --config .lychee.toml ./*.md  ./docs/**/*.md 2>&1 | grep -vP '\[(200|EXCLUDED)\]'
✗ [503] https://github.com/kubernetes/community/blob/master/contributors/guide/github-workflow.md#squash-commits | Failed: Network error: Service Unavailable
✗ [503] https://github.com/syself/cluster-api-provider-hetzner/blob/main/README.md#%EF%B8%8F-compatibility-with-cluster-api-and-kubernetes-versions | Failed: Network error: Service Unavailable
✗ [503] https://github.com/syself/cluster-api-provider-hetzner/blob/main/CONTRIBUTING.md | Failed: Network error: Service Unavailable

Issues found in 3 inputs. Find details below.

[./CONTRIBUTING.md]:
✗ [503] https://github.com/kubernetes/community/blob/master/contributors/guide/github-workflow.md#squash-commits | Failed: Network error: Service Unavailable

[docs/caph/01-getting-started/02-quickstart/02-management-cluster-setup.md]:
✗ [503] https://github.com/syself/cluster-api-provider-hetzner/blob/main/README.md#%EF%B8%8F-compatibility-with-cluster-api-and-kubernetes-versions | Failed: Network error: Service Unavailable

[docs/caph/04-developers/01-development-guide.md]:
✗ [503] https://github.com/syself/cluster-api-provider-hetzner/blob/main/CONTRIBUTING.md | Failed: Network error: Service Unavailable

🔍 121 Total (in 26s) ✅ 103 OK 🚫 3 Errors 💤 15 Excluded
💡 There were issues with GitHub URLs. You could try setting a GitHub token and running lychee again.
make: *** [Makefile:765: lint-links] Error 2
Error: Process completed with exit code 2.
```

We should cache that somehow. 

If no single markdown file has changed, then Lychee should be skipped.

For fixing above errors setting GITHUB_TOKEN should be enough.

--- a/.github/workflows/pr-lint.yml
+++ b/.github/workflows/pr-lint.yml
@@ -44,6 +44,8 @@ jobs:
       # if there's a diff then the workflow will exit here.
       - name: Run make verify
         run: make BUILD_IN_CONTAINER=false verify
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
       - name: Lint Golang Code
         run: make BUILD_IN_CONTAINER=false lint-golang-ci

--- a/Makefile
+++ b/Makefile
@@ -762,6 +762,7 @@ ifeq ($(BUILD_IN_CONTAINER),true)
 		$(BUILDER_IMAGE):$(BUILDER_IMAGE_VERSION) $@;
 else
 	@lychee --version
+	@if [ -z "$${GITHUB_TOKEN}" ]; then echo "GITHUB_TOKEN is not set"; exit 1; fi
 	lychee --verbose --config .lychee.toml ./*.md  ./docs/**/*.md 2>&1 | grep -vP '\[(200|EXCLUDED)\]'
 endif
 

